swingset/src/main/java/com/nqadmin/swingset/SSBaseComboBox.java

```diff
@@ -302,6 +303,25 @@ protected BaseGlazedModel() {
 		}
 	}
 
+	/**
+	 * A list item formatter that displays the mapping if the option is null.
+	 */
+	@SuppressWarnings("serial")
+	class ShowMappingIfNullOption extends SSListItemFormat {
+
+		@Override
+		protected void appendValue(StringBuffer _sb, int _elemIndex, SSListItem _listItem) {
+			if (optionModel.getOptionListItemElemIndex() == _elemIndex
+					&& getElem(_elemIndex, _listItem) == null) {
+				_sb.append("=== ")
+					.append(getElem(optionModel.getMappingListItemElemIndex(), _listItem).toString())
```

Comment:
This is the exception line. Since optionModel has already been used, I think the only NPE would be trying to do toString() because of a null return from
```
getElem(optionModel.getMappingListItemElemIndex(), _listItem)
```
and, in OptionMappingSwingModel, there's
```
	public int getMappingListItemElemIndex() {
		return KEY_IDX;
	}
```
So, IIRC, this means that not only is the displayed value null, but the key is also null. This could be some startup or initialization condition, unless it's really a null key. Either way this looks like a reasonable solution:
```
Object key = getElem(optionModel.getMappingListItemElemIndex(), _listItem);
_sb.append("=== ")
	.append(key != null ? key.toString() : null)
	.append(" ===");

```
to avoid the NPE, and see if there's any further issue.

Since I don't see any tests as part of this PR, I guess I just used the existing tests and probably walked through this code. If this doesn't work, or you don't want to try the fix, provide a test case and I'll be happy to update the PR. Something added in the Demo would be good.

Reply:
BTW, if you run this with JDK-17 it will say exactly what caused the NPE

Reply:
There is explicit key with value of null. But I do remember the combo boxes were updated to allow the user to remove a selection (empty selection which would set a null value in database). So the code probably is adding an option with null value as key to the list of user specified mappings.

Reply:
Yeah, that could be, maybe one of the demos will NPE. May want to special case null to leave it blank?

Guess that's for after it works...

Reply:
using the below code so that we are showing empty item that users can select (to deselect a value). This is solving the NPE issue. Will test the display of key.

Object key = getElem(optionModel.getMappingListItemElemIndex(), _listItem); 
_sb.append(key != null ? key.toString() : "");

Reply:
Tested with a combo which did not have the mapping needed for a key, the key was not displayed when the mapping is not present.

Reply:
Uh, too many negatives. Is this a good result? Sounds like it's a bad result.

In the code I think mapping means key, option means value.

Reply:
Yeah, not getting the expected result. The combo box is blank when there is no option/value to display. Expecting to see the mapping/key from the database.

Reply:
Weird, since we're pretty sure it's getting there because of the NPE fix (but that could have been for a different record). That would seem to indicate that
```
Object key = getElem(optionModel.getMappingListItemElemIndex(), _listItem);
```
it returning a null.

I'm trying to see how I can reproduce this with the demo, the demo sql is at the following files, if you have any ideas on how I can reproduce the issue

https://github.com/bpangburn/swingset/blob/82abc676eebda08508c77ef5b539235f72aa8bbc/swingset-demo/src/main/resources/mysql.swingset-demo-app.sql#L45
https://github.com/bpangburn/swingset/blob/82abc676eebda08508c77ef5b539235f72aa8bbc/swingset-demo/src/main/resources/mysql.swingset-demo-components.sql#L44

I'll fire up the demo and see if I can see something

Reply:
Those are the wrong pointers; they are the mysql versions, The normal ones are in the same directory. I think I can do it, I'll let you know.

Reply:
On the line

https://github.com/bpangburn/swingset/blob/82abc676eebda08508c77ef5b539235f72aa8bbc/swingset-demo/src/main/resources/suppliers_and_parts.sql#L61

I change 'Jones' to NULL and I see a blank, I think that's the problem. I wonder if a NULL is getting converted to blank when the records are getting read from the database.

Reply:
No. Sent the email via standard mail server.

Reply:
I had realized the 'Jones' thing was a red herring, when your mail came in.

With the change you suggested, I just saw two items (and the nullItem) in the combo box. So I tried
```
cmbPartColor.setOptions(new String[] { null, "Green", "Blue" });
```
And with that, here's what I see

![NullOption](https://user-images.githubusercontent.com/20450427/193322186-b2c24829-4e39-4cff-b01f-b2a3d56a3fc9.png)

Reply:
We need the ability to see the underlying database value without explicitly setting it in the options. If we know that we have to set, say code 0, we would set it with the right value. This request is to cover scenarios where 1. Developer forgot to add a key that is needed 2. Combo box items are based on some values in a database table but some how the value in the database (bound column) has some other value.

Reply:
That makes sense. I wonder where the record is being tossed out or ignored? 

The change you gave me
```
cmbPartColor.setOptions(new String[] { "Green", "Blue" }, new int[] {1,2});
```
covers case 1 I guess. 

How to reproduce case 2? Looks like the test might cover case 2 as well.

Reply:
Is this different behavior that pre-4 swingset? What used to happen in swingset 2/3?

Reply:
This is the same behaviour we had right from the beginning. Showing the underlying database value rather than blank would be an enhancement.

Reply:
Always good to know I didn't break anything... Guess my "make it behave the same" skills worked... Actually, this is probably in code I didn't really change.

I'm in the middle of something, should be able to track down the troublesome code over the next few days. Then we can discuss how to handle it. Maybe keep the mapping in the list and give the missing option a null value, then the new list item format code would kick in and display the mapping.

Reply:
I am guessing this needs a change in updateSSComponent of SSBaseComboBox. Maybe java would not allow us to do what we really want. We should not have that key as an option to choose when user tries to select a value. Should be used for display only.

Reply:
> We should not have that key as an option to choose

Why not? It's already in the data base, but I see what you mean. 

If it's in the combo, you can choose it; I guess the old behavior avoided choosing it :-) We could consider a flag that's part of the listItem that means can't be chosen, but that all sounds pretty horrible.

Isn't there some event when the combo chooses a value? At that point could ignore the choice and/or throw up an alert/warning.

Reply:
(there was a shift in requirements, found some time)

In `SSBaseComboBox` I added
```
+       boolean ADDITEM_HACK = true;
+       private int addItemHack(BaseModel<M,O,O2>.Remodel remodel, M _mapping) {
+               int index = -1;
+               if(_mapping != null) {
+                       if(ADDITEM_HACK) {
+                               remodel.add(_mapping, null);
+                               index = remodel.getMappings().indexOf(_mapping);
+                       }
+               }
+               return index;
+       }
```
and in `SetSelectedMapping(M _mapping)` I added
```
                        SSListItem item;
                        if (index != -1) {
                                item = remodel.get(index);
+                       } else if((index = addItemHack(remodel, _mapping)) != -1) {
+                               item = remodel.get(index);
                        } else {
                                // nullItem is either special first list item
```
And with the change you suggested to show the problem, when ex4 comes up

![NullOpt-fix](https://user-images.githubusercontent.com/20450427/193358984-5bf48023-4710-4047-bc2c-1ee47cdc066d.png)

Reply:
So does this support the developer, and is OK for production? 

I might display `*** 0 *** REPORT ISSUE` or some such.

Might want a property like `enableDiscoveredMapping` or something.

Reply:
In Example4, to verify that a custom list formatter could be done by the application, I've got
```
// SETUP THE COMBO BOX OPTIONS TO BE DISPLAYED AND THEIR CORRESPONDING VALUES
cmbPartColor.setListItemFormat(new SSListItemFormat() {
    @Override
    protected void appendValue(StringBuffer _sb, int _elemIndex, SSListItem _listItem) {
        if (cmbPartColor.getOptionFormatIndex() == _elemIndex
                && getElem(_elemIndex, _listItem) == null) {
            Object key = getElem(cmbPartColor.getMappingFormatIndex(), _listItem);
            _sb.append("*** ")
                    .append(key != null ? key.toString() : null)
                    .append(" *** REPORT ME");
        } else {
            super.appendValue(_sb, _elemIndex, _listItem);
        }
    }
    
});
//cmbPartColor.setOptions(new String[] { "Red", "Green", "Blue" });
cmbPartColor.setOptions(new String[] { "Green", "Blue" }, new int[] {1,2});
```
giving

![NullOpt-fix2](https://user-images.githubusercontent.com/20450427/193364305-1c11c4a2-5f67-49bd-912e-28a1073baf07.png)